herddb-core/src/main/java/herddb/core/TableManager.java

```diff
@@ -3529,6 +3531,46 @@ public boolean isStarted() {
         return started;
     }
 
+    @Override
+    public void validateAlterTable(Table table, StatementEvaluationContext context) throws StatementExecutionException {
+        List<String> columnsChangedFromNullToNotNull = new ArrayList<>();
+        for (Column c : this.table.columns) {
+            Column newColumnSpecs = table.getColumn(c.name);
+            if (newColumnSpecs == null) {
+                // dropped column
+                LOGGER.log(Level.INFO, "Table {0}.{1} dropping column {2}", new Object[]{table.tablespace, table.name, c.name});
+            } else if (newColumnSpecs.type == c.type) {
+                // no data type change
+            } else if (ColumnTypes.isNotNullToNullConversion(c.type, newColumnSpecs.type)) {
+                LOGGER.log(Level.INFO, "Table {0}.{1} making column {2} NULLABLE", new Object[]{table.tablespace, table.name, newColumnSpecs.name});
+            } else if (ColumnTypes.isNullToNotNullConversion(c.type, newColumnSpecs.type)) {
+                LOGGER.log(Level.INFO, "Table {0}.{1} making column {2} NOT NULL", new Object[]{table.tablespace, table.name, newColumnSpecs.name});
+                columnsChangedFromNullToNotNull.add(c.name);
+            }
+        }
+        for (final String column : columnsChangedFromNullToNotNull) {
+            LOGGER.log(Level.INFO, "Table {0}.{1} validating column {2}, check for NULL values", new Object[]{table.tablespace, table.name, column});
+            ScanStatement scan = new ScanStatement(this.table.tablespace,
+                    this.table, new Predicate() {
+                @Override
+                public boolean evaluate(Record record, StatementEvaluationContext context) throws StatementExecutionException {
+                    return record.getDataAccessor(table).get(column) == null;
+                }
+            });
+            // fast fail
+            scan.setLimits(new ScanLimitsImpl(1, 0));
+            boolean foundOneNull = false;
+            try (DataScanner scanner = this.scan(scan, context, null, false, false);) {
+                foundOneNull = scanner.hasNext();
+            } catch (DataScannerException err) {
+                throw new StatementExecutionException(err);
+            }
+            if (foundOneNull) {
+                throw new StatementExecutionException("Found a record in table " + table.name + " that contains a NULL value for column " + column + " ALTER command is not possible");
```

Comment:
I see this is where you are doing the null check on the column values. For a large table will this operation be a problem? This happens async?

Reply:
This is blocking.
The alter statement will complete only after this code has been fully executed

We aren't blocking the table explicitly so actually we could have problems if another client is modifying the table concurrently.

Blocking the table, like acquiring a global write (checkpoint) lock is doable but I did not do it because it should be implemented carefully.

The immediate need is to alter table schema while the application is stopped so this is the minimal change.

We should fix the concurrent access case in a follow up case.

Currently (before this patch) the implementation is already broken because if you try to change a column from not null to null able the alter command succeeds but it really does not have effect (the column stays nullable!)